rpc/wrtc_call_queue_mongodb.go

```diff
@@ -584,30 +592,61 @@ func (queue *mongoDBWebRTCCallQueue) operatorLivenessLoop() {
 				{webrtcOperatorHostsAnswererSizeField, sizes.Answerer},
 			})
 		}
-
-		result, err := queue.operatorsColl.UpdateOne(queue.cancelCtx, bson.D{{webrtcOperatorIDField, queue.operatorID}}, bson.D{
-			{
-				"$set",
-				bson.D{
-					{webrtcOperatorExpireAtField, time.Now().Add(operatorHeartbeatWindow)},
-					{webrtcOperatorHostsField, hostSizes},
+		updateCtx, cancel := context.WithTimeout(queue.cancelCtx, operatorHeartbeatWindow/3)
+		start := time.Now()
+		result, err := queue.operatorsColl.UpdateOne(
+			updateCtx,
+			bson.D{{webrtcOperatorIDField, queue.operatorID}},
+			bson.D{
+				{
+					"$set",
+					bson.D{
+						{webrtcOperatorExpireAtField, time.Now().Add(operatorHeartbeatWindow)},
+						{webrtcOperatorHostsField, hostSizes},
+					},
 				},
 			},
-		})
+			// There is a chance that multiple previous updates took too long or failed and the document expired.
+			// In these cases it is acceptable to upsert a new operator document.
+			options.Update().SetUpsert(true),
+		)
 		if err != nil {
 			reason := "context_canceled"
 			if !errors.Is(err, context.Canceled) {
-				//nolint:goconst
-				reason = "other"
+				switch {
+				case errors.Is(err, context.DeadlineExceeded):
+					reason = "deadline_exceeded"
+				default:
+					//nolint:goconst
+					reason = "other"
+				}
 				queue.logger.Errorw("failed to update operator document for self", "error", err)
 			}
 			operatorsCollUpdateFailures.Inc(queue.operatorID, reason)
-		} else if result.MatchedCount == 0 {
-			queue.logger.Errorw(
-				"no existing operator document found, could not update operator document",
-				"operator_id", queue.operatorID,
+		} else if result.UpsertedCount == 1 {
+			queue.logger.Infow("no existing operator document found, inserted new operator document")
+			// not technically a failure, but interesting to track
+			operatorsCollUpserts.Inc(queue.operatorID)
+		}
+		cancel()
+
+		// Answerer liveness is determined by whether the operator document is updated.
+		// If updates to the operator document fail continuously, the answerers are
+		// not live as callers will not be able to make connections to the answerers connected
+		// to this queue. Therefore, if update failed, do not run the onAnswererLiveness func.
+		// This errcheck is done above as well, but pulled the continue statement out
+		// into a separate block for readability.
```

Comment:
maybe a bit debatable whether we should do this, but I think this makes everything more consistent. But you can definitely still get into a case as pointed out by the TODO where answerers can be connected to the signaler but not reachable.

Reply:
I think this a good thing to do.